lib/pry/history.rb

```diff
@@ -144,5 +146,19 @@ def history_file
     def history_file_path
       File.expand_path(@file_path || Pry.config.history.file)
     end
+
+    def remove_store_and_prepare_to_restore_irb_history
+      restore_irb_history_at_session_end Readline::HISTORY.to_a
+      clear_readline
+    end
+
+    def restore_irb_history_at_session_end(irb_history)
+      return if Pry.hooks.hook_exists?(:after_session, :restore_irb_history)
+
+      Pry.hooks.add_hook(:after_session, :restore_irb_history) do
+        Readline::HISTORY.shift until Readline::HISTORY.empty?
+        Readline::HISTORY << irb_history.shift until irb_history.empty?
+      end
```

Comment:
This seems a bit fishy. You only add the hook once, but the hook closes over the set of lines to restore. Doesn't that mean that if we actually enter and leave Pry twice, the lines restored at the end of the second session will be the ones that were in `HISTORY` before the first session?